File: ios/FlowKit/Sources/FlowKit/Connection/ConnectionLifetime.swift
import Foundation

/// Shared by real-host connections for one app mode. A cancelled lifetime can never become active again.
public final class ConnectionLifetime: @unchecked Sendable {
    private let lock = NSLock()
    private var active = true
    private var handlers: [UUID: @Sendable () -> Void] = [:]

    public init() {}
    public var isActive: Bool { lock.withLock { active } }

    /// Serializes transport enqueueing with cancellation. The closure must not suspend.
    func whileActive(_ body: () throws -> Void) rethrows -> Bool {
        try lock.withLock {
            guard active else { return false }
            try body()
            return true
        }
    }

    func onCancel(_ id: UUID, _ body: @escaping @Sendable () -> Void) {
        let cancelled = lock.withLock {
            if active { handlers[id] = body }
            return !active
        }
        if cancelled { body() }
    }

    func remove(_ id: UUID) { _ = lock.withLock { handlers.removeValue(forKey: id) } }

    public func cancel() {
        let callbacks = lock.withLock {
            active = false
            let callbacks = Array(handlers.values)
            handlers.removeAll()
            return callbacks
        }
        for callback in callbacks { callback() }
    }
}

File: android/core/src/main/kotlin/com/inferenceaftermath/remotly/core/connection/ConnectionLifetime.kt
package com.inferenceaftermath.remotly.core.connection

/** Shared by real-host connections for one app mode. Cancellation is permanent: a later mode gets a new lifetime. */
class ConnectionLifetime {
    private val lock = Any()
    private var active = true
    private val handlers = mutableMapOf<Any, () -> Unit>()

    val isActive: Boolean get() = synchronized(lock) { active }

    /** Serializes handing bytes to the transport with cancellation; never suspend inside [block]. */
    internal fun whileActive(block: () -> Unit): Boolean = synchronized(lock) {
        if (!active) false else { block(); true }
    }

    internal fun onCancel(key: Any, block: () -> Unit) {
        val cancelled = synchronized(lock) {
            if (active) handlers[key] = block
            !active
        }
        if (cancelled) block()
    }

    internal fun remove(key: Any) { synchronized(lock) { handlers.remove(key) } }

    fun cancel() {
        val callbacks = synchronized(lock) {
            if (!active) return
            active = false
            handlers.values.toList().also { handlers.clear() }
        }
        // Stop callbacks take each connection's lock; never call them while holding this lock.
        callbacks.forEach { it() }
    }
}
